x/gov/types/msgs.go

```diff
@@ -36,23 +36,39 @@ func (msg MsgSubmitProposal) Type() string { return TypeMsgSubmitProposal }
 
 // ValidateBasic implements Msg
 func (msg MsgSubmitProposal) ValidateBasic() sdk.Error {
-	if msg.Content == nil {
+	content, err := msg.Content()
+	if err != nil {
+		return sdk.ConvertError(err)
+	}
+	if content == nil {
 		return ErrInvalidProposalContent(DefaultCodespace, "missing content")
 	}
-	if msg.Proposer.Empty() {
-		return sdk.ErrInvalidAddress(msg.Proposer.String())
+	var proposer sdk.AccAddress
+	proposer, err = msg.Proposer()
+	if err != nil {
```

Comment:
Huh, these `err != nil` checks are quite annoying. Otherwise, this use of a cap'n proto native struct doesn't seem so much worse than a normal struct (dev-ux-wise).

This is unavoidable, as Unmarshal usually returns an error for parse error, and we cannot actually validate all fields until we read them (as there is no parsing phase). I am wondering on how to make this simpler for the developer, short of the [go 2 proposal for `check` keyword](https://go.googlesource.com/proposal/+/master/design/go2draft-error-handling-overview.md). Note that if cross-language support is a central concern, this issue is irrelevant in eg. rust or javascript. Javascript will throw and exception, and Rust will return a result, which can be short-circuited by simply appending `?` (even easier than `check`).

If we do the example like `CoinsEToCoins`, then that conversion becomes a parse function, which may return an error (naturally), but the rest of the usage of the struct must not worry about any errors, as all elements have already been parsed. In the sdk use case, it is highly unlikely we will ignore fields and never parse them (unlike some streaming api use cases), so the "parse function" (also mentioned above) seems like a good way to handle the ergonomics.